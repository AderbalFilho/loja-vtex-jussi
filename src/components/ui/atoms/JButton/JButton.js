import React from 'react';
import classNames from 'classnames/bind';
import PropTypes from 'prop-types';

import './JButton.scss';

function JButton({
  backgroundColor,
  border,
  children,
  classes,
  color,
  radius,
  tag,
  ...rest
}) {
  const className = classNames(
    'j-button',
    `j-button--back-${backgroundColor}`,
    `j-button--border-${border}`,
    `j-button--${color}`,
    `j-button--rad-${radius}`,
    tag === 'a' && `j-button--link`,
    classes
  );
  const HtmlTag = tag;

  return (
    <HtmlTag className={className} {...rest}>
      {children}
    </HtmlTag>
  );
}

JButton.defaultProps = {
  backgroundColor: 'transparent',
  border: 'black',
  color: 'black',
  radius: 'normal',
  tag: 'button',
};

JButton.propTypes = {
  backgroundColor: PropTypes.oneOf(['transparent', 'green']),
  border: PropTypes.oneOf(['black', 'pink', 'green']),
  children: PropTypes.node,
  classes: PropTypes.string,
  color: PropTypes.oneOf(['black', 'pink', 'green']),
  radius: PropTypes.oneOf(['normal', 'circle']),
  tag: PropTypes.oneOf(['button', 'a']),
};

export default JButton;
