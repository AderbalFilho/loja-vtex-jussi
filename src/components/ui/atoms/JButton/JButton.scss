@import 'styles/mixin';

.j-button {
  @extend %text-paragraph;

  padding: 12px 24px;
}

.j-button--back-transparent {
  background-color: transparent;
}

.j-button--back-green {
  background-color: var(--green-jussi);
}

.j-button--border-black {
  border: 1px solid var(--black);
}

.j-button--border-pink {
  border: 1px solid var(--pink-jussi);
}

.j-button--border-green {
  border: 1px solid var(--green-jussi);
}

.j-button--black {
  color: var(--black);
}

.j-button--pink {
  color: var(--pink-jussi);
}

.j-button--green {
  color: var(--green-jussi);
}

.j-button--rad-normal {
  border-radius: 6px;
}

.j-button--rad-circle {
  border-radius: 24px;
}

.j-button--link {
  text-decoration: none;
}
