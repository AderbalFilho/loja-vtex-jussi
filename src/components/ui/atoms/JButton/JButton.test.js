import React from 'react';
import TestRenderer from 'react-test-renderer';
import { render } from '@testing-library/react';

import JButton from './JButton';

describe('JButton', () => {
  describe('change background-color', () => {
    it('renders transparent when no backgroundColor is passed', () => {
      const { container } = render(<JButton>Jüssi</JButton>);
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--back-transparent');
    });

    it('renders green', () => {
      const { container } = render(
        <JButton backgroundColor="green">Jüssi</JButton>
      );
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--back-green');
    });

    it('renders transparent', () => {
      const { container } = render(
        <JButton backgroundColor="transparent">Jüssi</JButton>
      );
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--back-transparent');
    });
  });

  describe('change border color', () => {
    it('renders black when no border is passed', () => {
      const { container } = render(<JButton>Jüssi</JButton>);
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--border-black');
    });

    it('renders pink', () => {
      const { container } = render(<JButton border="pink">Jüssi</JButton>);
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--border-pink');
    });

    it('renders green', () => {
      const { container } = render(<JButton border="green">Jüssi</JButton>);
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--border-green');
    });

    it('renders black', () => {
      const { container } = render(<JButton border="black">Jüssi</JButton>);
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--border-black');
    });
  });

  describe('change color', () => {
    it('renders black when no color is passed', () => {
      const { container } = render(<JButton>Jüssi</JButton>);
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--black');
    });

    it('renders pink', () => {
      const { container } = render(<JButton color="pink">Jüssi</JButton>);
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--pink');
    });

    it('renders black', () => {
      const { container } = render(<JButton color="black">Jüssi</JButton>);
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--black');
    });
  });

  describe('change radius', () => {
    it('renders normal when no radius is passed', () => {
      const { container } = render(<JButton>Jüssi</JButton>);
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--rad-normal');
    });

    it('renders circle', () => {
      const { container } = render(<JButton radius="circle">Jüssi</JButton>);
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--rad-circle');
    });

    it('renders normal', () => {
      const { container } = render(<JButton radius="normal">Jüssi</JButton>);
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--rad-normal');
    });
  });

  describe('change tag', () => {
    it('matches snapshot with button when no tag is passed', () => {
      const testRenderer = TestRenderer.create(<JButton>Jüssi</JButton>);
      const jsonComponent = testRenderer.toJSON();

      expect(jsonComponent).toMatchSnapshot();
    });

    it('renders a', () => {
      const { container } = render(
        <JButton tag="a" href="https://jussi.com.br/">
          Jüssi
        </JButton>
      );
      const button = container.querySelector('.j-button');

      expect(button.classList).toContain('j-button--link');
    });

    it('matches snapshot with a', () => {
      const testRenderer = TestRenderer.create(
        <JButton tag="a" href="https://jussi.com.br/">
          Jüssi
        </JButton>
      );
      const jsonComponent = testRenderer.toJSON();

      expect(jsonComponent).toMatchSnapshot();
    });

    it('matches snapshot with button', () => {
      const testRenderer = TestRenderer.create(
        <JButton tag="button">Jüssi</JButton>
      );
      const jsonComponent = testRenderer.toJSON();

      expect(jsonComponent).toMatchSnapshot();
    });
  });

  it('renders children', () => {
    const { container } = render(<JButton>Jüssi</JButton>);
    const button = container.querySelector('.j-button');

    expect(button.textContent).toBe('Jüssi');
  });

  it('renders classes', () => {
    const { container } = render(
      <JButton classes="button-offer">Jüssi</JButton>
    );
    const button = container.querySelector('.j-button');

    expect(button.classList).toContain('button-offer');
  });

  it('renders rest', () => {
    const { container } = render(
      <JButton tag="a" href="/buy">
        Jüssi
      </JButton>
    );
    const button = container.querySelector('.j-button');

    expect(button.getAttribute('href')).toBe('/buy');
  });
});
