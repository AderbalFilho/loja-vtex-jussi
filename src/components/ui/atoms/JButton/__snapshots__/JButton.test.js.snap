// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JButton change tag matches snapshot with a 1`] = `
<a
  className="j-button j-button--back-transparent j-button--border-black j-button--black j-button--rad-normal j-button--link"
  href="https://jussi.com.br/"
>
  Jüssi
</a>
`;

exports[`JButton change tag matches snapshot with button 1`] = `
<button
  className="j-button j-button--back-transparent j-button--border-black j-button--black j-button--rad-normal"
>
  Jüssi
</button>
`;

exports[`JButton change tag matches snapshot with button when no tag is passed 1`] = `
<button
  className="j-button j-button--back-transparent j-button--border-black j-button--black j-button--rad-normal"
>
  Jüssi
</button>
`;
